### PR TITLE
Upd/Generics have landed

### LINKED ISSUE #2181: Use generics for morph caches
We cache containers, container ids, eACLs and IR nodes. Casting all of them to `ttlNetCache` is a bit ugly, we could use generics after min go version is bumped.

### PR DESCRIPTION
**Pull request by notimetoname**
Closes #2181. Blocked by #2342.

### CODE REVIEW

**Comment by roman-khimov** [comment]
See #2319 for things to fix when dropping 1.17 support. There is #2181, but I'm not so immediately sure about that.

**Comment by roman-khimov** [comment]
Can we split it to #2319 and #2181 parts?

**Comment by notimetoname** [comment]
#2342 is done. Lets have generic fun.

File: cmd/neofs-node/cache.go [hunk 1]
```diff
@@ -53,47 +53,45 @@ func newNetworkTTLCache(sz int, ttl time.Duration, netRdr netValueReader) *ttlNe
...
 
-func (c *ttlNetCache) remove(key interface{}) {
+func (c *ttlNetCache[K, V]) remove(key K) {
 	c.cache.Remove(key)
 }
 
 // entity that provides LRU cache interface.
 type lruNetCache struct {
-	cache *lru.Cache
+	cache *lru.Cache[uint64, *netmapSDK.NetMap]
```
Comment (roman-khimov) [inline cmd/neofs-node/cache.go:87]: As much as I'm skeptical about generics, this probably makes things more readable.

File: cmd/neofs-node/cache.go [hunk 2]
```diff
@@ -15,32 +15,32 @@ import (
...
 
 // entity that provides TTL cache interface.
-type ttlNetCache struct {
+type ttlNetCache[K comparable, V any] struct {
 	ttl time.Duration
 
 	sz int
 
-	cache *lru.Cache
+	cache *lru.Cache[K, *valueWithTime[V]]
```
Comment (roman-khimov) [inline cmd/neofs-node/cache.go:33]: While this maybe not so much.

File: cmd/neofs-node/cache.go [hunk 3]
```diff
@@ -125,73 +123,63 @@ func (c *lruNetCache) get(key interface{}) (interface{}, error) {
...
 // that implements container storage.
-type ttlContainerStorage ttlNetCache
+type ttlContainerStorage struct {
+	tc ttlNetCache[cid.ID, *container.Container]
+}
 
 func newCachedContainerStorage(v container.Source, ttl time.Duration) *ttlContainerStorage {
 	const containerCacheSize = 100
 
-	lruCnrCache := newNetworkTTLCache(containerCacheSize, ttl, func(key interface{}) (interface{}, error) {
```
Comment (roman-khimov) [inline cmd/neofs-node/cache.go:133]: Changes like this are good as well.

**Review by roman-khimov — COMMENTED** [review]
Can't say I'm impressed with the diffstat though (`+137 −159`), my expectations were somewhat higher. Something tells me we'll need to revise these caches anyway.

Please add #2181 fix/close somewhere.

**Comment by notimetoname** [comment]
> Please add https://github.com/nspcc-dev/neofs-node/issues/2181 fix/close somewhere.

@roman-khimov, added to PR's description. It is 8 commits now, not sure which of them should have that link (or all?).

> Can't say I'm impressed with the diffstat though (+137 −159), my expectations were somewhat higher.

In fact, mine was too. But i like that we do not have casts (ok, "assertions") in that code more, i find that parentheses dance (4+ in one expression) hard to read. Also, i would say that they just do their work, and if it can be done with a small amount of code it should be done. In other words, i would say that our caches are too complex, so generics do not make it shorter where nothing relates to "generic code".

Anyway, i think, LRU caches will be updated to `v2` somehow, so the issue says to update it, that PR updates it. Any other refactor should be considered in a separate way (can create an issue about that).

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**